### PR TITLE
[CIT-565] tedge connect test c8y/az

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes
'tedge connct c8y/az --test` should return an error on fail to check the connection.
Also before calling check connection, it should check if the device is connected or not.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: tedge/src/cli/connect/command.rs [hunk 1]
```diff
@@ -213,7 +238,7 @@ async fn check_device_status_c8y(port: u16) -> Result<DeviceStatus, ConnectError
                     println!("... No response. If the device is new, it's normal to get no response in the first try.");
                 } else {
                     println!("... No response. ");
-                    return Ok(DeviceStatus::Unknown);
+                    return Err(ConnectError::ConnectionCheckError);
```
Comment (didier-wenzek) [inline tedge/src/cli/connect/command.rs:241]: This PR should be just that: the heart of the fix is here.

File: tedge/src/cli/connect/command.rs [hunk 2]
```diff
@@ -165,6 +174,22 @@ impl ConnectCommand {
...
+            .join(TEDGE_BRIDGE_CONF_DIR_PATH)
+            .join(br_config.config_file);
+
+        match Path::new(&bridge_conf_path).exists() {
+            // If we find the bridge config file we remove it
+            // and carry on to see if we need to restart mosquitto.
+            true => true,
+            false => false,
+        }
+    }
```
Comment (didier-wenzek) [inline tedge/src/cli/connect/command.rs:192]: I have a negative feeling about this kind of check. That code is supposed to have created that file. If not it should have returned an error *before* running that test that is useless.

Furthermore:
* The comment is misleading: do we really remove the file it exists?
* A match true => true, false => false, is useless.

File: tedge/src/cli/connect/error.rs [hunk 3]
```diff
@@ -38,4 +38,10 @@ pub enum ConnectError {
 
     #[error(transparent)]
     ConfigLoadError(#[from] tedge_config::TEdgeConfigError),
+
+    #[error("Connection check failed")]
+    ConnectionCheckError,
+
+    #[error("Device is not connected to {cloud} cloud")]
+    BridgeFileDoesNotExist { cloud: String },
```
Comment (didier-wenzek) [inline tedge/src/cli/connect/error.rs:46]: Please remove this error. This file is not provided by user, but supposed to be created by the `tedge connect` command.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
This PR should focus on the bug fix: the exit code is no more 1 on error.

File: tedge/src/cli/connect/command.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply tedge/src/cli/connect/command.rs:192]: I removed those comments, and fixed the code logic, and simplified the code. The whole purpose here is to check locally if the device is connected to the cloud or not before doing the actual check_connection. With the existing code even if the device is not connected still goes ahead and publishes the messages on the topic, if we do this check before we can avoid this.

File: tedge/src/cli/connect/error.rs (hunk 3 above)
Comment (PradeepKiruvale) [reply tedge/src/cli/connect/error.rs]: I feel its necessary, I renamed it to "DeviceNotConnected"

**Review by albinsuresh — APPROVED** [review]
Changes look fine to me. I guess we need a test for `tedge connect c8y --test` if it doesn't exist already to prevent similar failures in future.

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**